lib/delayed_job/threaded_worker.rb

```diff
@@ -0,0 +1,104 @@
+require 'socket'
+
+class ThreadedWorker < Delayed::Worker
+  def initialize(thread_count, options={})
+    super(options)
+    @thread_count = thread_count
+    @threads = []
+    @stop_signal = false
+    @name_prefix = options[:worker_name] || 'worker'
+  end
+
+  def start
+    trap_signals
+
+    say "Starting multi-threaded job worker with #{@thread_count} threads"
+
+    @thread_count.times do |i|
+      thread_name = generate_thread_name(i + 1)
+      @threads << Thread.new do
+        thread_name(thread_name)
+        threaded_work_off
+      end
+    end
+
+    @threads.each(&:join) # Wait for threads to finish
+  end
+
+  def name
+    Thread.current[:name] || @name
+  end
+
+  def name=(name)
+    # Set the instance variable for compatibility && ensure thread-local storage is also updated
+    @name = name
+    Thread.current[:name] = name
+  end
+
+  private
+
+  def generate_thread_name(thread_index)
+    if @name
+      "#{@name}-thread:#{thread_index}"
+    else
+      "#{@name_prefix}-host:#{Socket.gethostname} pid:#{Process.pid} thread:#{thread_index}"
+    end
+  end
+
+  def thread_name(name)
+    # Store the name in thread-local storage && Set the name in the instance variable for compatibility
+    Thread.current[:name] = name
+    self.name = name
+  end
+
+  def stop_signal?
+    @stop_signal
+  end
+
+  def threaded_work_off
+    retry_attempts = 0
+
+    until stop_signal?
+      begin
+        runtime = Benchmark.realtime do
+          @result = work_off(100) # Attempt to process one job
+        end
+
+        count = @result[0] + @result[1]
+
+        if count.zero?
+          sleep(self.class.sleep_delay) unless stop_signal?
+        else
+          say sprintf("#{count} jobs processed at %.4f j/s, %d failed", count / runtime, @result.last)
+        end
+      rescue StandardError => e
+        say "Thread #{name} encountered an error: #{e.message}. Restarting thread..."
+        retry_attempts += 1
+        if retry_attempts >= 5
+          say "Thread #{name} has failed #{retry_attempts} times. Exiting to prevent infinite loop."
```

Comment:
**question** Just to understand here, with our current DelayedWorker, if there is an exception in the thread it will exit the process and monit will start it back up again.

With this implementation if one of our threads fails 5 times it will return, but the thread nor process will be restarted unless *all* threads error and the `join` above completes and we exit the process?

I'm not sure in what scenarios multiple exceptions are raised, but its possible over time our actual running thread count will reduce from `thread_count` downwards until finally all threads error and monit restarts the process. Is that correct?

Reply:
This could indeed become a problem. I think it would be better to use a thread pool (with `Concurrent::ThreadPoolExecutor`). I have a first working implementation but the tests still need some work. I'll convert this PR to a draft for now.

Reply:
We decided to simplify the implementation by removing the thread pool and retry mechanisms.
In case a thread errors we will log the error message trigger a graceful shutdown of 30 seconds and raise an error. Then monit can restart the whole process.
This is probably a edge case as delayed job already catches all errors when executing jobs: https://github.com/collectiveidea/delayed_job/blob/master/lib/delayed/worker.rb#L241-L244